### PR TITLE
chore: Add security scanning with FOSSA and Snyk

### PR DESCRIPTION
**Pull request by leordev**
cc @ALRubinger

### CODE REVIEW

File: .github/workflows/security.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+jobs:
+  snyk:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.head_ref }}
+
```
Comment (alecthomas) [inline .github/workflows/security.yml:29]: You'll need to activate Hermit to get the correct versions of toolchains:

```yaml
- uses: cashapp/activate-hermit@v1
```

Comment (alecthomas) [reply .github/workflows/security.yml:29]: Aha already done!

**Comment by leordev** [comment]
@alecthomas @worstell any ideas of how to solve [this error](https://github.com/TBD54566975/ftl/actions/runs/7174908241/job/19537255510?pr=722#step:4:403)?

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.2.0:tree (default-cli) on project ftl: Cannot build project dependency graph: Could not resolve nor collect following dependencies: [xyz.block:ftl-runtime:jar:1.0-SNAPSHOT (compile)]: Could not resolve dependencies for project ftl:ftl:pom:1.0-SNAPSHOT: The following artifacts could not be resolved: xyz.block:ftl-runtime:jar:1.0-SNAPSHOT (absent): Could not find artifact xyz.block:ftl-runtime:jar:1.0-SNAPSHOT -> [Help 1]

**Comment by leordev** [comment]
> The examples won't compile without having first installed the FTL libraries. Running `mvn install` from the root should do it.

Does executing `mvn install` from untrusted forked PRs can execute any malicious code? Or it will always install benevolent packages? 

Eg: say something add some malicious dependency package, is it possible it execute any malicious binary code? Or it's a simple pull of maven jars? (@ALRubinger maybe you know the answer here?)

Asking because this workflow trigger event `pull_request_target` has access to our exposed secrets, in this case SNYK_TOKEN and FOSSA_API_KEY. (Security reference [here](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/))

**Comment by alecthomas** [comment]
Actions from untrusted forked PRs aren't executed by default. An admin has to approve them.

**Comment by ALRubinger** [comment]
Yep @leordev exactly as @alecthomas says. Relies on the project leads to review the deps being brought in (and the code in the PRs!) to ensure there's no env var sniffing there which would expose our secrets. Good on ya @leordev for flagging so the FTL team is aware.

**Comment by alecthomas** [comment]
I'm a bit curious why a security scan needs to build anything though, to be honest. That seems like a very odd thing to be doing during a scan.

**Comment by leordev** [comment]
> I'm a bit curious why a security scan needs to build anything though, to be honest. That seems like a very odd thing to be doing during a scan.

Same here... We can observe from the CI logs that FOSSA didnt need to install anything, whereas Snyk is complaining about it... I tried to run the checks locally without a java/go environment and confirmed above too... FOSSA is able to scan... and Snyk complains about mvn and go cmds not found....

I'm thinking about adding the `mvn install` in the CI step to unblock this and trust that external PRs will going to be reviewed before approving their workflow runs then. WDYT?

**Comment by ALRubinger** [comment]
I guessing it's because the `mvn` build is what runs the dependency resolution step, leading to how we get the full dependency tree in the Maven Effective POM. The full dep tree isn't available from the `pom.xml` alone.

**Comment by alecthomas** [comment]
I'm going to close this. Feel free to reopen once resuming work.

**Pull request closed without merging**